llama.cpp

```diff
@@ -12771,8 +12773,10 @@ void llama_grammar_accept_token(struct llama_context * ctx, struct llama_grammar
     // Note terminating 0 in decoded string
     const auto   decoded     = decode_utf8(piece, grammar->partial_utf8);
     const auto & code_points = decoded.first;
+    std::vector<std::vector<const llama_grammar_element *>> tmp_new_stacks;
     for (auto it = code_points.begin(), end = code_points.end() - 1; it != end; ++it) {
-        grammar->stacks = llama_grammar_accept(grammar->rules, grammar->stacks, *it);
+        llama_grammar_accept(grammar->rules, grammar->stacks, *it, tmp_new_stacks);
+        tmp_new_stacks.swap(grammar->stacks);
```

Comment:
Is this better than saying `grammar->stacks = tmp_new_stacks;`? Because `new_stacks` is `.clear()`'d on 11921, it seems like we don't need to save its value here, and we could save a small step (?).

Mainly though, the recursive nature of the swap here was making my eyes cross when trying to follow exactly what this change was doing and how the contents of `grammar->stacks` and `tmp_new_stacks` were ping-ponging back and forth in this loop, so getting rid of the `.swap()` might make it a bit easier to read as well?

Reply:
FWIW, I tried making this change (into a local `grammar-speedup4` branch), and it didn't significantly improve things, but it wasn't slower, and I think the code is a bit more readable:

```
Benchmark 1: ./main \
        -mu https://huggingface.co/TheBloke/phi-2-GGUF/resolve/main/phi-2.Q4_K_M.gguf \
        --grammar-file json_numbers.grammar \
        -p "List of 20 integers starting from 0" \
        --seed 12344 (branch = grammar-speedup4)
  Time (mean ± σ):     12.586 s ±  0.698 s    [User: 8.488 s, System: 1.799 s]
  Range (min … max):   12.012 s … 13.726 s    5 runs

Benchmark 2: ./main \
        -mu https://huggingface.co/TheBloke/phi-2-GGUF/resolve/main/phi-2.Q4_K_M.gguf \
        --grammar-file json_numbers.grammar \
        -p "List of 20 integers starting from 0" \
        --seed 12344 (branch = grammar-speedup3)
  Time (mean ± σ):     12.904 s ±  0.854 s    [User: 8.583 s, System: 1.954 s]
  Range (min … max):   11.846 s … 13.963 s    5 runs

Summary
  ./main \
        -mu https://huggingface.co/TheBloke/phi-2-GGUF/resolve/main/phi-2.Q4_K_M.gguf \
        --grammar-file json_numbers.grammar \
        -p "List of 20 integers starting from 0" \
        --seed 12344 (branch = grammar-speedup4) ran
    1.03 ± 0.09 times faster than ./main \
        -mu https://huggingface.co/TheBloke/phi-2-GGUF/resolve/main/phi-2.Q4_K_M.gguf \
        --grammar-file json_numbers.grammar \
        -p "List of 20 integers starting from 0" \
        --seed 12344 (branch = grammar-speedup3)
```

Reply:
Heh, turns out my eyes-crossing swap wasn't even making things faster, removed it / looks simpler thanks.